features/storage/nvstore/source/nvstore.cpp

```diff
@@ -891,8 +890,8 @@ int NVStore::init()
 
         // We have a non valid master record, in a non-empty area. Just erase the area.
         if ((!valid) || (key != master_record_key)) {
-            os_ret = flash_erase_area(area);
-            MBED_ASSERT(!os_ret);
+            ret = flash_erase_area(area);
+            MBED_ASSERT(!ret);
```

Comment:
It is slightly worrying that the errors due to HW problems are detected only on debug-builds. If the remaining code survives from storage error, what is the point of assert here?

Reply:
Do we expect hardware problems to be indicated by error returns from `flash` functions though?

It's quite possible that the only error returns would be for parameter errors like address being out of range, and we only find out flash failure by actually trying to use it and finding it doesn't work.

But I would agree that if these functions can indicate hardware problems, there should be paths to deal with that. Asserts should be reserved for catching should-be-impossible design errors.

Reply:
Tend to agree here. :-) But, well out of scope for this PR in a way.

Reply:
> Tend to agree here. :-) But, well out of scope for this PR in a way.

Please create a tracking issue for this.

Reply:
Done, IOTSTOR-832

Reply:
@kjbracey-arm at least some drivers seem to notify HW/driver problems, not just user errors:
https://github.com/ARMmbed/mbed-os/blob/cc94690363d66f7437b532eddddc7e6afea5036b/drivers/FlashIAP.cpp#L204
->
https://github.com/ARMmbed/mbed-os/blob/d2e9fde70114a5e395a532fc5234699aafa92784/targets/TARGET_Cypress/TARGET_PSOC6/flash_api.c#L41-L50